hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -4180,22 +4185,54 @@ func (r *HostedClusterReconciler) reconcileAWSSubnets(ctx context.Context, creat
 	return nil
 }
 
-func isUpgradeable(hcluster *hyperv1.HostedCluster) (bool, string, error) {
-	if hcluster.Status.Version != nil && hcluster.Status.Version.Desired.Image != hcluster.Spec.Release.Image {
-		upgradeable := meta.FindStatusCondition(hcluster.Status.Conditions, string(hyperv1.ClusterVersionUpgradeable))
-		if upgradeable != nil && upgradeable.Status == metav1.ConditionFalse {
-			releaseImage, exists := hcluster.Annotations[hyperv1.ForceUpgradeToAnnotation]
-			if !exists {
-				return true, "", fmt.Errorf("cluster version is not upgradeable")
-			} else if releaseImage != hcluster.Spec.Release.Image {
-				return true, "", fmt.Errorf("cluster version is not upgradeable, force annotation is present but does not match desired release image")
-			} else {
-				return true, "cluster version is not upgradeable, upgrade is forced by annotation", nil
-			}
-		}
+func (r *HostedClusterReconciler) lookupReleaseImage(ctx context.Context, hcluster *hyperv1.HostedCluster) (*releaseinfo.ReleaseImage, error) {
+	var pullSecret corev1.Secret
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: hcluster.Namespace, Name: hcluster.Spec.PullSecret.Name}, &pullSecret); err != nil {
+		return nil, fmt.Errorf("failed to get pull secret: %w", err)
+	}
+	pullSecretBytes, ok := pullSecret.Data[corev1.DockerConfigJsonKey]
+	if !ok {
+		return nil, fmt.Errorf("expected %s key in pull secret", corev1.DockerConfigJsonKey)
+	}
+	return r.ReleaseProvider.Lookup(ctx, hcluster.Spec.Release.Image, pullSecretBytes)
+}
+
+// isUpgradeable returns
+// 1) bool indicating whether the HostedCluster is upgrading
+// 2) non-error message about the condition of the upgrade
+// 3) error indicating that the upgrade is not allowed or we were not able to determine
+func isUpgradeable(hcluster *hyperv1.HostedCluster, releaseImage *releaseinfo.ReleaseImage) (bool, string, error) {
+	if hcluster.Status.Version == nil || hcluster.Status.Version.Desired.Image == hcluster.Spec.Release.Image {
```

Comment:
Reviewer note: I inverted some of the logic here so we can just return early and make it clearer.